addon/index.d.ts

```diff
@@ -0,0 +1,940 @@
+ * operations.
+ *
+ * You can also define an
+ * <a href="/docs/encapsulated-task">Encapsulated Task</a>
+ * by passing in an object that defined a `perform` generator
+ * method.
+ *
+ * The following Component defines a task called `myTask` that,
+ * when performed, prints a message to the console, sleeps for 1 second,
+ * prints a final message to the console, and then completes.
+ *
+ * ```js
+ * import { task, timeout } from 'ember-concurrency';
+ * export default Component.extend({
+ *   myTask: task(function * () {
+ *     console.log("Pausing for a second...");
+ *     yield timeout(1000);
+ *     console.log("Done!");
+ *   })
+ * });
+ * ```
+ *
+ * ```hbs
+ * <button {{action myTask.perform}}>Perform Task</button>
+ * ```
+ *
+ * By default, tasks have no concurrency constraints
+ * (multiple instances of a task can be running at the same time)
+ * but much of a power of tasks lies in proper usage of Task Modifiers
+ * that you can apply to a task.
+ *
+ * @param taskFn A generator function backing the task or an encapsulated task descriptor object with a `perform` generator method.
+ */
+export function task<T extends TaskFunction<any, any[]>>(taskFn: T):
+  TaskProperty<TaskFunctionReturnType<T>, TaskFunctionArgs<T>>;
+export function task<T extends EncapsulatedTaskDescriptor<any, any[]>>(taskFn: T):
+  TaskProperty<EncapsulatedTaskDescriptorReturnType<T>, EncapsulatedTaskDescriptorArgs<T>>;
+
+/**
+ * "Task Groups" provide a means for applying
+ * task modifiers to groups of tasks. Once a {@linkcode Task} is declared
+ * as part of a group task, modifiers like `drop()` or `restartable()`
+ * will no longer affect the individual `Task`. Instead those
+ * modifiers can be applied to the entire group.
+ *
+ * ```js
+ * import { task, taskGroup } from 'ember-concurrency';
+ *
+ * export default Controller.extend({
+ *   chores: taskGroup().drop(),
+ *
+ *   mowLawn:       task(taskFn).group('chores'),
+ *   doDishes:      task(taskFn).group('chores'),
+ *   changeDiapers: task(taskFn).group('chores')
+ * });
+ * ```
+ *
+ * @returns {TaskGroupProperty}
+ */
+export function taskGroup<T>(): TaskGroupProperty<T>;
+
+/**
+ * A cancelation-aware variant of [Promise.all](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all).
+ * The normal version of a `Promise.all` just returns a regular, uncancelable
+ * Promise. The `ember-concurrency` variant of `all()` has the following
+ * additional behavior:
+ *
+ * - if the task that `yield`ed `all()` is canceled, any of the
+ *   {@linkcode TaskInstance}s passed in to `all` will be canceled
+ * - if any of the {@linkcode TaskInstance}s (or regular promises) passed in reject (or
+ *   are canceled), all of the other unfinished `TaskInstance`s will
+ *   be automatically canceled.
+ *
+ * [Check out the "Awaiting Multiple Child Tasks example"](/docs/examples/joining-tasks)
+ */
+export function all<T extends readonly unknown[] | readonly [unknown]>(
```

Comment:
One note and one question here—

- If we want to support tuples (it's nice!), we should add support for some further (if limited) number of tuples, `[unknown, unknown]`, `[unknown, unknown, unknown]`, etc.—I'd suggest going out to four or five, as that's what we've generally done with Ember's own types.
- Why the `readonly`?

Reply:
As far as I can tell, the `T1, T2, T3, ...` approach is obsolete. This is based on some of the newer built-in types in TS (I think this is based on the latest `Promise.all`).

There are tests for this, but basically the `unknown[]`  covers the other possibilities, while the `...[K in keyof T]...` in the return type will match everything nicely. I think `[unknown]` is needed as a special case otherwise it will get typed as `unknown[]` which is less precise and has some undesirable consequences. 

As far as `readonly`, admittedly it's mostly just because I copied it, but I believe that's because `readonly` is the narrowest possible thing here and it's what we actually require, so it's good to be precise, otherwise, if someone has an array that's marked readonly for some reasons, they won't be able to pass it in.

In general, I think it's a good idea to mark array-based APIs as `readonly` where possible for that reason.

Reply:
Interesting. I will need to dig in a bit and figure out what value the `| [unknown]` adds, because I can [confirm][playground] that it infers correctly with *just* `unknown[]`.

In any case, I’m happy enough to leave it given it’s copied from `Promise` directly. 👍 Thanks for clarifying!

[playground]: https://www.staging-typescript.org/play?#code/C4TwDgpgBAShDOB7ANgNwgEwDwBUB8UAvFDlBAB7AQB2G8UACgE6IC2AlvBADLsDWELO2oAzCE0YEA-IygAuEgG4AUMowQAxsgCGTaCICu1DcHaJqUbcmS4ylGnSh7tGc8hBQjfaogDu1AG0AXTwACmUoSKhUKwMEBRwAGmUASgVmNk5BAG8oAFpnV2p3KACAaShhKAEQRBESIIU4JDRMXHKQqABfPBVVdS1daGQIYEsFAPhgJmEAcyCVAZ09KBGxgCMFKZnqWeDFzWXh0agNCe25xKhqA1Z18QXVK2RQ7RTlZ9D198+NFKA